### PR TITLE
avoid storing ship_info pointers in data structures

### PR DESCRIPTION
**Pull request by Goober5000**
For collections that we want to make dynamic, we should not store pointers in data structures because these pointers will be invalidated if the collection is resized.  This PR takes care of all `ship_info` pointers stored in data structures in the codebase.

Since `WarpEffect` stored pointers to objects and ships (and warp params, which is already dynamic), this PR handles those as well.  There are, of course, more places where ship and object pointers are stored that will eventually need to be handled.

### CODE REVIEW

**Review by JohnAFernandez — CHANGES_REQUESTED** [review]
One quick thing

**Comment by Baezon** [comment]
I was actually thinking about recently changing some `weapon_info_index` fields into just `weapon_info*`. *Technically*, being in a vector they are "dynamic", but morally, they aren't. We parse them once, and then the vector is never changed again, so theoretically pointers should still be valid? While parsing perhaps, the vector may be changing size, is that the worry?

File: code/scripting/api/objs/cockpit_display.cpp [hunk 1]
```diff
@@ -8,19 +8,18 @@
...
 	if (!this->isValid())
 		return NULL;
 
-	return &m_sip->displays[m_display_num];
+	return &Ship_info[m_ship_info_idx].displays[m_display_num];
 }
-bool cockpit_disp_info_h::isValid() {
-	if (m_sip == NULL)
+bool cockpit_disp_info_h::isValid() const {
+	if (m_ship_info_idx < 0)
```
[the remark this answers is not in the document]
Comment (Goober5000) [reply code/scripting/api/objs/cockpit_display.cpp]: Sure.

**Comment by Goober5000** [comment]
> I was actually thinking about recently changing some `weapon_info_index` fields into just `weapon_info*`. _Technically_, being in a vector they are "dynamic", but morally, they aren't. We parse them once, and then the vector is never changed again, so theoretically pointers should still be valid? While parsing perhaps, the vector may be changing size, is that the worry?

I see where you're coming from, and in practice the pointer is likely to be safe, at least at the current time.  But I think this is one of those "err on the safe side" situations.  I can think of a few reasons to store ship and weapon info indexes rather than pointers:
1. If we should store indexes for most pointer types, then we might as well be consistent for info types as well.
2. A pointer might work today, but we could run the risk of it failing in the future for some unknown reason.  See #5785 for an example of a recent fix.  (The waypoint didn't cause problems in FSO, but did cause a crash in FRED.)
3. If someone ever implements a way to reload tables on the fly (while the engine is running), this would cause the info pointers to be invalidated.

**Comment by EatThePath** [comment]
To add to the above, if the scp_pool effort comes to fruition it's use requires we store indexes instead of pointers. Anything already storing indexes can be a lot easier to convert.

**Review by JohnAFernandez — APPROVED** [review]

**Pull request merged**